File: python/behavioral/observer/MainObserver.py
import random
from abc import ABC, abstractmethod

class Observer(ABC):
    @abstractmethod
    def update(self, data: str):
        pass


class ConcreteObserver(Observer):
    def __init__(self):
        self.__i = random.random()
    def update(self, data: str):
        print(data + " " + str(self.__i))

class Subject(ABC):
    def attach(self, observer: Observer):
        pass
    def detach(self, observer: Observer):
        pass
    def notifyObservers(self):
        pass

class ConcreteSubject(Subject):
    def __init__(self):
        self.__subjectState: str = None
        self.__observers: set[Observer] = set()

    def attach(self, observer: Observer):
        self.__observers.add(observer)

    def detach(self, observer: Observer):
        self.__observers.remove(observer)

    def notifyObservers(self):
        for observer in self.__observers:
            observer.update(self.__subjectState)

    def getSubjectState(self):
        return self.__subjectState

    def setSubjectState(self, subjectState: str):
        self.__subjectState = subjectState


def main():
    observer1 = ConcreteObserver()
    observer2 = ConcreteObserver()
    concreteSubject = ConcreteSubject()
    concreteSubject.attach(observer1)
    concreteSubject.attach(observer2)
    concreteSubject.setSubjectState("Hello")
    concreteSubject.notifyObservers()
    concreteSubject.setSubjectState("Message")
    concreteSubject.notifyObservers()

main()

File: kotlin/src/main/kotlin/behavioral/observer/MainObserver.kt
package behavioral.observer

import kotlin.random.Random

internal interface Observer {
    fun update(data: String)
}

internal class ConcreteObserver : Observer {
    private val i: Int = Random(1).nextInt()
    override fun update(data: String) {
        println("$data $i")
    }
}

internal interface Subject {
    fun attach(observer: Observer)
    fun detach(observer: Observer)
    fun notifyObservers()
}

internal class ConcreteSubject : Subject {
    lateinit var subjectState: String
    private val observers: MutableSet<Observer> = mutableSetOf()
    override fun attach(observer: Observer) {
        observers.add(observer)
    }

    override fun detach(observer: Observer) {
        observers.remove(observer)
    }

    override fun notifyObservers() {
        for (observer in observers) {
            observer.update(subjectState)
        }
    }
}

fun main(args: Array<String>) {
    val observer1: Observer = ConcreteObserver()
    val observer2: Observer = ConcreteObserver()
    val concreteSubject = ConcreteSubject()
    concreteSubject.attach(observer1)
    concreteSubject.attach(observer2)
    concreteSubject.subjectState = "Hello"
    concreteSubject.notifyObservers()
}